Code/GraphMol/MarvinParse/MarvinDefs.cpp

```diff
@@ -63,11 +53,570 @@ ptree MarvinMolBase::toPtree() const {
 }
 
 void MarvinMolBase::addSgroupsToPtree(ptree &out) const {
-  for (auto sgroup : sgroups) {
+  for (auto &sgroup : sgroups) {
     out.add_child("molecule", sgroup->toPtree());
   }
 }
 
+bool getCleanDouble(std::string strToParse, double &outDouble) {
+  if (boost::algorithm::trim_copy(strToParse) !=
+      strToParse) {  // should be no white space
+    return false;
+  }
+  try {
+    outDouble = boost::lexical_cast<double>(strToParse);
+  } catch (const std::exception &e) {
+    return false;
+  }
+
+  return true;
+}
+
+bool getCleanInt(std::string strToParse, int &outInt) {
+  if (boost::algorithm::trim_copy(strToParse) !=
+      strToParse) {  // should be no white space
+    return false;
+  }
+  try {
+    outInt = boost::lexical_cast<int>(strToParse);
+  } catch (const std::exception &e) {
+    return false;
+  }
+
+  return true;
+}
+
+void MarvinMolBase::parseAtomsAndBonds(ptree &molTree) {
+  if (this->hasAtomBondBlocks()) {
+    boost::property_tree::ptree atomArray = molTree.get_child("atomArray");
+
+    // there are two types of atom arrays:
+    // <atomArray atomID="a1 a2 a3 a4 a5 a6 a7 a8 a9 a10 a11" elementType="C
+    // C C C C C Cl C N O O" formalCharge="0 0 0 0 0 0 0 0 1 0 -1"
+    // lonePair="0 0 0 0 0 0 3 0 0 2 3" x2="-4.3334 -5.6670 -5.6670 -4.3334
+    // -2.9997 -2.9997 -4.3335 -1.6660 -7.0007 -1.6660 -0.3323" y2="1.7693
+    // 0.9993 -0.5409 -1.3109 -0.5409 0.9993 3.3093 -1.3109 -1.3109 -2.8509
+    // -0.5410"></atomArray>
+    //  AND
+    // <atomArray>
+    //       <atom id="a1" elementType="C" x2="-9.4583" y2="1.9358"
+    //       mrvStereoGroup="and1"/> <atom id="a2" elementType="C"
+    //       x2="-10.7921" y2="1.1658"/> <atom id="a3" elementType="C"
+    //       x2="-10.7921" y2="-0.3744"/> <atom id="a8" elementType="O"
+    //       x2="-12.1257" y2="-1.1444" lonePair="2"/>
+    //   </atomArray>
+
+    // See which one we have
+
+    std::string atomID =
+        atomArray.get<std::string>("<xmlattr>.atomID", "UseLongForm");
+    if (atomID == "UseLongForm") {
+      // long form - each atom on a line
+
+      BOOST_FOREACH (boost::property_tree::ptree::value_type &v,
+                     molTree.get_child("atomArray")) {
+        if (v.first != "atom") {
+          continue;
+        }
+
+        auto *mrvAtom = new MarvinAtom();
+        this->pushOwnedAtom(mrvAtom);
```

Comment:
I'm really wondering about that pattern where you add the atoms in two vectors always

Reply:
THis has all changed.  I don't always add the atoms in both vectors.  Each molecule and sgroup has its own list of atoms - only the top parent mol has the Owned atoms now